src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultBoundAttributeDescriptorBuilder.cs

```diff
@@ -94,26 +87,19 @@ public override void BindAttributeParameter(Action<BoundAttributeParameterDescri
 
     public BoundAttributeDescriptor Build()
     {
-        var diagnostics = Validate();
-        if (_diagnostics != null)
-        {
-            diagnostics ??= new();
-            diagnostics.UnionWith(_diagnostics);
-        }
+        using var _ = HashSetPool<RazorDiagnostic>.GetPooledObject(out var diagnostics);
```

Comment:
💡 Pooled collections perform best when code avoids taking an object from a pool until such time as it actually needs to use it. We've had to fix bugs in the Roslyn compiler related to early leases from the object pool leading to exceptionally high GC rates (allocation rates similar to without the pool, but allocation costs higher due to cost of the pool itself).

Reply:
That's fair. I'd considered that initially but didn't love the ergonomics of the code needed to achieve that. You're definitely right though, and I'll go ahead and make the change.

Reply:
I decided on a decent abstraction for this that actually makes the code a bit cleaner. :smile: